cluster/vm-nocloud.yaml

```diff
@@ -0,0 +1,29 @@
+metadata:
+  name: testvm-nocloud
+apiVersion: kubevirt.io/v1alpha1
+kind: VM
+spec:
+  domain:
+    devices:
+      disks:
+      - type: ContainerRegistryDisk:v1alpha
+        source:
+          name: kubevirt/cirros-registry-disk-demo:devel
+        target:
+          dev: vda
+      interfaces:
+      - source:
+          network: default
+        type: network
+    memory:
+      unit: MB
+      value: 64
+    os:
+      type:
+        os: hvm
+    type: qemu
+  cloudInit:
```

Comment:
Looking at this example, I'd really go with the disk approach.

If we go with the disk approach, then we don't clobber the spec, and the user got explicit conroll (and this is what VM API is about, explicit controll) over where the disk is located.
VMConfig could be the place where cloud-init is exposed as a`feature`.

Thus:
```
disk:
- type: CloudInit
  mode: noCloud # needed?
  userData: …
  target:
    dev: vda
```

Reply:
> Looking at this example, I'd really go with the disk approach.

I disagree. The 'nocloud' datasource happens to use a disk. There are other cloud-init data sources that will not.  

I'd rather not see this configuration data as something that is intermingled with the VM spec. These are logically two separate concepts.  The VM spec defines what the domain looks like, The cloud-init spec defines the cloud-init datasource method used to inject configuration into the the VM.  The latter could involve injecting a disk into the VM, or it could involve talking to a meta-data server. 

> f we go with the disk approach, then we don't clobber the spec, and the user got explicit conroll (and this is what VM API is about, explicit controll) over where the disk is located.
VMConfig could be the place where cloud-init is exposed as a feature.

I'm doubtful that we'll ever care about where a disk is placed on a vm that uses cloud-init and the nocloud data source.

Can you give an example of when disk placement matters?

Reply:
the disk naming needs to be predictable to the guest with or without cloud-init. The only way to make sure it fits within other disks is to make it part of the spec. The actual content can be separate, but there should be a disk device referring to it in the spec section

Reply:
> the disk naming needs to be predictable to the guest with or without cloud-init. The only way to make sure it fits within other disks is to make it part of the spec. The actual content can be separate, but there should be a disk device referring to it in the spec section

This coordination already happens, but it is the inverse of what you're proposing. The CloudInit spec's 'nocloud' structure takes in the 'diskTarget' name that is used when injecting the disk into the domain xml.

Reply:
> I disagree. The 'nocloud' datasource happens to use a disk. There are other cloud-init data sources that will not.

But there is only one noCloud source, and this is disk based iiuic.



> I'd rather not see this configuration data as something that is intermingled with the VM spec. These are logically two separate concepts.


> I'm doubtful that we'll ever care about where a disk is placed on a vm that uses cloud-init and the nocloud data source.

We do need to care about this, for some operating system (Windows) it is important what disk it is, and what it's label is - so we do need to have control.

By making it a disk, a consumer (like us) has the explciti control over where the disk is placed.



> This coordination already happens, but it is the inverse of what you're proposing. The CloudInit spec's 'nocloud' structure takes in the 'diskTarget' name that is used when injecting the disk into the domain xml.

The disk target field is not recognized by all OSes we can not depend on it. Bus/device addresses are relevant. These could be implicitly defined by the ordering of the devices - but the disk target is not reliable.

Reply:
> The disk target field is not recognized by all OSes we can not depend on it. Bus/device addresses are relevant. These could be implicitly defined by the ordering of the devices - but the disk target is not reliable.

@fabiand I'm fairly certain the diskTarget is going to be recognized by all OSes that care about the nocloud data source.... or cloud-init at all for that matter.  

I know the Windows use case where we have to care about disk ordering and need explicit control. That use case isn't applicable here. cloud-init is not a windows thing.

Reply:
I believe there is an integration with Openstack for https://cloudbase.it/cloudbase-init/ ?

Reply:
@davidvossel it's not only about recognizing the nocloud disk (I trust you it works today, though I still have a vivid memory of really crappy cloud-init versions in RHEL 6) but equally about not screwing up other disks. You do not want the other disks to "shift".  If the disk is explicitly in domain spec you it's more clear how it affects the rest of the VM

Reply:
We were discussing this yesterday offline, and it will go to the disk. Either by referencing the could-init data in the disk source section, or by putting the whole cloud-init section there ( (up to david). Both should be good enough, right?

Reply:
sure, both is fine. keeping it separate(referencing) might be better for other non-disk ways of specifying cloudinit, like openstack metadata service. But it doesn't really matter much indeed